pkg/operator/controller.go

```diff
@@ -299,7 +299,12 @@ func (c *Controller) sync() error {
 				updatedCR.ObjectMeta = cr.ObjectMeta
 			}
 			if specChanged {
+				// FIXME: Here be dragons. The operator can
+				// accidentally lose user-provided
+				// configuration.
+				managementState := updatedCR.Spec.ManagementState
 				updatedCR.Spec = cr.Spec
+				updatedCR.Spec.ManagementState = managementState
```

Comment:
why are we ever stomping what is in etcd w/ the values from the cached copy?  we don't own the spec values, the user does

Reply:
because the operator populates a lot of fields in spec during bootstrap, and we don't want to loose this data. If the operator decides to update `spec.storage` concurrently with the user, the operator wins. Once the registry config is bootstrapped, the operator shouldn't touch spec anymore and new user changes won't be stomped.
It shouldn't be a problem for real customers (as they should get their cluster once the config is bootstrapped), but it is a problem for tests as they change config too fast.

Reply:
isn't it in danger of happening any time a sync(resync?) is happening where the user has made a change to the object, but the cache isn't updated yet?

(yes this is probably rare, but the only time we'd have an update conflict is if exactly that happened, right?)

if the cache and etcd are out of sync, maybe it's better to abandon the sync attempt and instead update the cache (which will trigger syncing)?

Reply:
The simplified code is

```go
cr := getConfigFromCache()
prevCR := cr.DeepCopy()
reconcile(&cr)
if metadataOrSpecChanged(prevCR, cr) {
  saveConfig(cr)
}
```

Normally `reconcile` logic doesn't change `cr`, so we don't get into this `saveConfig` path. It'll happen only if the cached config has empty fields. And normally customers get their cluster after the operator has finished generation of the config, i.e. the operator will already have the "good" config in its cache.

It can be a problem if somebody tries to change the config and deletes something from the config. In this case the operator will switch back into bootstrapping mode and can stomp user changes.

It can become a bigger problem if we make the image registry optional, i.e. when somebody can write a script:
1. enable the operator,
2. change its config.